data-pipeline-flink/notification-engine/src/main/scala/org/sunbird/dp/notification/function/IncompleteCourseReminderEmailNotification.scala

```diff
@@ -125,174 +146,87 @@ class IncompleteCourseReminderEmailNotification(courseConfig:NotificationEngineC
       }
     })
   }
 
   def getAndSetUserEmail(userCourseMap: java.util.Map[String, UserCourseProgressDetails]): Unit = {
-    val userIds: java.util.List[String] = new java.util.ArrayList[String]()
-    var userDetailsListRow: java.util.List[Row] = new java.util.ArrayList[Row]()
-    val isDeleted = false
-    userIds.addAll(userCourseMap.keySet())
-    val query = QueryBuilder.select().column(courseConfig.EMAIL).from(courseConfig.dbSunbirdKeyspace, courseConfig.EXCLUDE_USER_EMAILS).allowFiltering().toString
-    val excludeEmailsRow = cassandraUtil.find(query)
-    val excludeEmailsList: java.util.List[Any] = new java.util.ArrayList[Any]()
-    excludeEmailsRow.forEach(email=>excludeEmailsList.add(email.getString(0)))
-    userIds.forEach(id=>{
-      val queryForUserDetails = QueryBuilder.select().column(courseConfig.ID).column(courseConfig.PROFILE_DETAILS_KEY).from(courseConfig.dbSunbirdKeyspace, courseConfig.TABLE_USER)
-        .where(QueryBuilder.eq("id", id))
-        .and(QueryBuilder.eq("isDeleted", isDeleted))
-        .and(QueryBuilder.eq("status", 1)).allowFiltering().toString
-      val rowData = cassandraUtil.find(queryForUserDetails)
-      userDetailsListRow.addAll(rowData)
-    })
-    userDetailsListRow.forEach(userDetails=>{
-      try {
-        if (userDetails.getString(courseConfig.PROFILE_DETAILS_KEY) != null) {
-          val profileDetails: String = userDetails.getString(courseConfig.PROFILE_DETAILS_KEY)
-          val profileDetailsMap: java.util.HashMap[String, Any] =new ObjectMapper().readValue(profileDetails, classOf[java.util.HashMap[String, Any]])
-          val personalDetailsMap: java.util.HashMap[String, Any] =profileDetailsMap.get(courseConfig.PERSONAL_DETAILS_KEY).asInstanceOf[util.HashMap[String,Any]]
-          if(MapUtils.isNotEmpty(personalDetailsMap)){
-            val primaryEmail: String = personalDetailsMap.get(courseConfig.PRIMARY_EMAIL).toString
-            if(StringUtils.isNotEmpty(primaryEmail) && !excludeEmailsList.contains(primaryEmail)){
-              val userId = userDetails.getString(courseConfig.ID)
+    try {
+      val userIds: java.util.List[String] = new java.util.ArrayList[String]()
+      userIds.addAll(userCourseMap.keySet())
+      val query = QueryBuilder.select().column(courseConfig.EMAIL).from(courseConfig.dbSunbirdKeyspace, courseConfig.EXCLUDE_USER_EMAILS).allowFiltering().toString
+      val excludeEmailsRow = cassandraUtil.find(query)
+      val excludeEmailsList: java.util.List[Any] = new java.util.ArrayList[Any]()
+      excludeEmailsRow.forEach(email => excludeEmailsList.add(email.getString(0)))
+      logger.info("exclude emails " + excludeEmailsList.size())
+      val limit: Int = 45
+      for (i <- 0 until userIds.size by limit) {
+        val userIdSubList = userIds.subList(i, Math.min(userIds.size(), i + limit))
+        var response = new util.ArrayList[util.HashMap[String, Any]]()
+        val query: BoolQueryBuilder = QueryBuilders.boolQuery()
```

Comment:
so this logic is like create a query with batch for elastic search . This was the enhancement of the previous code while previously I do one by one user Id and get the email id but now it will do 45 userId details with one hit and also instead of db fetch used elastic search

Reply:
same removed query variable